### PR TITLE
Fix #26 & #12: report line and column numbers on left of code and visually differentiate them

### LINKED ISSUE #26: Show line numbers to the left of code extract
Currently we show line numbers at the end of the file name:

```
[1] path/to/File.scala:21:
    not found: value foo
      foo()
      ^
```

This makes it difficult to quickly find the line number because we have to scan across the entire file name looking for the end of the line. It's a small problem for simple path hierarchies but gets bigger as projects get more complex package hierarchies.

I propose copying Elm and Rust and putting the line number immediately to the left of the code extract:

```
[E1] path/to/File.scala:
     not found: value foo
     L21:  Bar.foo()
               ^
```

From the familiarity perspective as well, people are more used to seeing line numbers to the left of their code.

### PR DESCRIPTION
**Pull request by yawaramin**
The string manipulation logic should be good, but I've never developed or tested an sbt plugin before, so would appreciate some help with the testing.

### CODE REVIEW

File: src/main/contraband-scala/sbt/errorssummary/ReporterConfig.scala [hunk 1]
```diff
@@ -99,47 +60,8 @@ final class ReporterConfig private (val colors: Boolean,
...
-                       warningColor,
-                       infoColor,
-                       debugColor,
-                       sourcePathColor,
-                       errorIdColor)
+  
+  def apply(): ReporterConfig = new ReporterConfig(true, true, false, scala.Console.RED, scala.Console.YELLOW, scala.Console.CYAN, scala.Console.WHITE, scala.Console.UNDERLINED, scala.Console.BLUE)
+  def apply(colors: Boolean, shortenPaths: Boolean, columnNumbers: Boolean): ReporterConfig = new ReporterConfig(colors, shortenPaths, columnNumbers, scala.Console.RED, scala.Console.YELLOW, scala.Console.CYAN, scala.Console.WHITE, scala.Console.UNDERLINED, scala.Console.BLUE)
+  def apply(colors: Boolean, shortenPaths: Boolean, columnNumbers: Boolean, errorColor: String, warningColor: String, infoColor: String, debugColor: String, sourcePathColor: String, errorIdColor: String): ReporterConfig = new ReporterConfig(colors, shortenPaths, columnNumbers, errorColor, warningColor, infoColor, debugColor, sourcePathColor, errorIdColor)
 }
```
Comment (Duhemm) [inline src/main/contraband-scala/sbt/errorssummary/ReporterConfig.scala]: Is this you IDE formatting the code? Could you revert this change please? This code gets generated, and I format it with Scalafmt.

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala [hunk 2]
```diff
@@ -64,6 +64,11 @@ private class ConciseReporter(logger: Logger,
             log(line)
           }
       }
+
+    logger.info("Legend:")
+    logger.info("Ln: line number n")
+    logger.info("Cn: column number n")
+    logger.info("En: error number n")
```
Comment (Duhemm) [inline src/main/scala/sbt/errorssummary/ConciseReporter.scala]: That'll make things much clearer. I have 2 suggestions:
 - Could you make it span on a single line? `Ln: line number / Cn: column number / En: ...`
 - Could you add a setting `showLegen` which defaults to `true`? You just need to edit this file: https://github.com/Duhemm/sbt-errors-summary/blob/master/src/main/contraband/ReporterConfig.contra. After editing, running `compile` in sbt will regenerate the code. (This PR shows how to add a setting: #28)

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala [hunk 3]
```diff
@@ -152,15 +157,25 @@ private class ConciseReporter(logger: Logger,
...
+          })
+
+        val position = colored(config.sourcePathColor, f)
+        val lineContent =
+          Option(problem.position.lineContent)
+            .filter(_.nonEmpty)
+            .map(lineCol.+)
+
+        val pointer = problem.position.pointerSpace.map { sp =>
+          "%s%s^".format(" " * lineCol.length, sp)
```
Comment (Duhemm) [inline src/main/scala/sbt/errorssummary/ConciseReporter.scala]: `lineCol` also contains the formatting content, so it's length is longer than what's displayed:

```scala
scala> import scala.Console._
import scala.Console._

scala> val s = RED + "foo" + RESET
s: String = foo

scala> s.length
res0: Int = 12
```

I think that the best solution would be to use `prefixed` and use `lineCol` as prefix. This way, you don't need to worry about alignment.

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala (hunk 2 above)
Comment (Duhemm) [reply src/main/scala/sbt/errorssummary/ConciseReporter.scala]: What do you think of having a second line that shows the different meanings of the colors? (Simply showing error with the colour for errors, warning with the color for warnings, etc.)

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala [hunk 4]
```diff
@@ -152,15 +157,25 @@ private class ConciseReporter(logger: Logger,
...
-          s"""${colored(config.sourcePathColor, f)}:$showLine$showCol"""
-        val lineContent = Some(problem.position.lineContent).filter(_.nonEmpty)
-        val pointer     = problem.position.pointerSpace.map(sp => s"$sp^")
-        val prefix      = s"${extraSpace(problem.severity)}[${problem.id}] "
+        val lineCol =
+          "%s%s".format(line.fold("") { l =>
+            colored(colorFor(problem), "L%d:".format(l))
+          }, offset.filter(_ => config.columnNumbers).fold("") { c =>
+            "C%d:".format(c + 1)
+          })
```
Comment (Duhemm) [inline src/main/scala/sbt/errorssummary/ConciseReporter.scala]: Any reason to prefer `format` over string interpolation? I would prefer to keep string interpolation if you don't mind.

File: src/test/scala/sbt/errorssummary/BasicConciseReporterSpec.scala [hunk 5]
```diff
@@ -91,23 +91,22 @@ class BasicConciseReporterSpec
...
-      (problems, messages) =>
-        problems should have length 1
+    collectMessagesFor(code, configWithShortenedPaths) { (problems, messages) =>
+      problems should have length 1
 
-        messages should have length 2
-        val (_, msg) = messages.head
-        val lines    = msg.split(EOL)
-        lines(0) shouldBe expectedText
+      messages should have length 2
```
Comment (Duhemm) [inline src/test/scala/sbt/errorssummary/BasicConciseReporterSpec.scala]: `messages` contains all the lines of output. No that you display the legend, it should be longer.

**Review by Duhemm — CHANGES_REQUESTED** [review]
Thanks! The new errors are certainly much clearer 👏🏻

To run the tests, you first need to run `setupTests` in sbt. Then, you can run the tests for Scala 2.10 with `test2106:test`, for 2.11 with `test21111:test` and 2.12 with `test2122:test`. Just look at what `.travis.yml` does 😉 

I don't think that you'll need to modify the `test-api` or `test-compiler` projects. If you do, you'll need to run `clean` and `setupTests` again.

Let me know if I can help and thanks again 🎉!

File: src/main/contraband-scala/sbt/errorssummary/ReporterConfig.scala (hunk 1 above)
Comment (yawaramin) [reply src/main/contraband-scala/sbt/errorssummary/ReporterConfig.scala]: Yup, it's the default generated code without IDE or Scalafmt formatting. I have IDEA set up with a plugin to run Scalafmt only on files that I edit. I'll reformat to match what you have.

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala (hunk 2 above)
Comment (yawaramin) [reply src/main/scala/sbt/errorssummary/ConciseReporter.scala]: Sure, I can make the legend a single line and add a setting. If you don't mind, perhaps we can design the enhanced (coloured) legend as a future issue/PR?

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala (hunk 4 above)
Comment (yawaramin) [reply src/main/scala/sbt/errorssummary/ConciseReporter.scala]: I find format strings a bit easier to understand, but no worries, I can switch to interpolation.

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala (hunk 3 above)
Comment (yawaramin) [reply src/main/scala/sbt/errorssummary/ConciseReporter.scala]: Ah, good point, and excellent suggestion. Will do.

File: src/test/scala/sbt/errorssummary/BasicConciseReporterSpec.scala (hunk 5 above)
Comment (yawaramin) [reply src/test/scala/sbt/errorssummary/BasicConciseReporterSpec.scala]: Oops, will make the change.

Comment (yawaramin) [reply src/test/scala/sbt/errorssummary/BasicConciseReporterSpec.scala]: Btw, it seems that all the tests are checking for `messages` length. Perhaps it's better for only one test (or one set of tests) to do the length checking?

**Comment by yawaramin** [comment]
OK, I believe that covers all the points except the coloured legend.

File: src/main/scala/sbt/errorssummary/ConciseReporter.scala (hunk 2 above)
Comment (Duhemm) [reply src/main/scala/sbt/errorssummary/ConciseReporter.scala]: Sure, no problem.

**Review by Duhemm — APPROVED** [review]
LGTM, thanks you very much!
I'll release that later today.

**Comment by olafurpg** [comment]
Thank you @yawaramin for this addition, I love seeing the line numbers on the left of the code!

**Pull request merged**